horovod/common/operations.cc

```diff
@@ -1049,7 +1065,48 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
         if (timeline.Initialized()) {
           RECORD_EVENT(entries, event_queue, NCCL_BCAST, stream)
         }
-      } else {
+      }
+
+      else if (horovod_global.use_dgc) {
+        auto &config = horovod_global.dgc_config;
+        auto &state  = horovod_global.dgc_state;
+
+        // DGC initialization
+        if (!config.configured) {
+          config.stream = stream;
+          config.nccl_comm = nccl_comm;
+          config.mpi_comm = horovod_global.mpi_comm;
+          config.global_num_gpus = horovod_global.size;
+          config.global_gpu_rank = horovod_global.rank;
+          config.local_comm      = horovod_global.local_comm;
+          config.local_num_gpus  = horovod_global.local_size;
+          config.local_gpu_rank  = horovod_global.local_rank;
+          config.cross_comm      = horovod_global.cross_comm;
+          config.global_num_nodes = horovod_global.cross_size;
+          config.global_node_rank = horovod_global.cross_rank;
+          config.configured = true;
+          state.layer_offset_bytes.clear();
+          state.step_counters.clear();
+          state.memory_table.clear();
+        }
+        config.context = entries[0].context;
+        config.device  = entries[0].device;
+
+        // Get names and sizes of each layer, in order
+        std::vector<std::pair<std::string, uint64_t> > layers;
+        layers.clear();
```

Comment:
Why do you clear it as soon as you make the vector.

Instead consider reserving that same size as entries: 
```diff
-    layers.clear();
+    layers.reserve(entries.size());
```